src/System.Management.Automation/CoreCLR/CorePsPlatform.cs

```diff
@@ -602,7 +604,190 @@ internal static int NonWindowsGetProcessParentPid(int pid)
+            }
+
+            string filename = startInfo.FileName;
+            string[] argv = ParseArgv(startInfo);
+            string[] envp = new string[0];
+            string cwd = !string.IsNullOrWhiteSpace(startInfo.WorkingDirectory) ? startInfo.WorkingDirectory : null;
+
+            // Invoke the shim fork/execve routine.  It will create pipes for all requested
+            // redirects, fork a child process, map the pipe ends onto the appropriate stdin/stdout/stderr
+            // descriptors, and execve to execute the requested process.  The shim implementation
+            // is used to fork/execve as executing managed code in a forked process is not safe (only
+            // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
+            int childPid, stdinFd, stdoutFd, stderrFd;
+            CreateProcess(
+                filename, argv, envp, cwd,
+                startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                SUPPRESS_PROCESS_SIGINT,    // Create SSH process to ignore SIGINT signals
+                out childPid,
+                out stdinFd, out stdoutFd, out stderrFd);
+
+            Debug.Assert(childPid >= 0);
+
+            // Configure the parent's ends of the redirection streams.
+            // We use UTF8 encoding without BOM by-default(instead of Console encoding as on Windows)
+            // as there is no good way to get this information from the native layer
+            // and we do not want to take dependency on Console contract.
+            if (startInfo.RedirectStandardInput)
+            {
+                Debug.Assert(stdinFd >= 0);
+                standardInput = new StreamWriter(OpenStream(stdinFd, FileAccess.Write),
+                    s_utf8NoBom, StreamBufferSize)
+                { AutoFlush = true };
+            }
+            if (startInfo.RedirectStandardOutput)
+            {
+                Debug.Assert(stdoutFd >= 0);
+                standardOutput = new StreamReader(OpenStream(stdoutFd, FileAccess.Read),
+                    startInfo.StandardOutputEncoding ?? s_utf8NoBom, true, StreamBufferSize);
+            }
+            if (startInfo.RedirectStandardError)
+            {
+                Debug.Assert(stderrFd >= 0);
+                standardError = new StreamReader(OpenStream(stderrFd, FileAccess.Read),
+                    startInfo.StandardErrorEncoding ?? s_utf8NoBom, true, StreamBufferSize);
+            }
+
+            return childPid;
+        }
+
+        /// <summary>Opens a stream around the specified file descriptor and with the specified access.</summary>
+        /// <param name="fd">The file descriptor.</param>
+        /// <param name="access">The access mode.</param>
+        /// <returns>The opened stream.</returns>
+        private static FileStream OpenStream(int fd, FileAccess access)
+        {
+            Debug.Assert(fd >= 0);
```

Comment:
Sure, but I do see the point of DotNet CoreFx developers who originated this code.  The assert condition is pretty explanatory...